benchmark/benchmark.py

```diff
@@ -237,6 +238,81 @@ def _get_ec2_instance_id() -> Optional[str]:
 
     return instance_id
 
+class ResourceMonitoring():
+    def __init__(self, with_bwm: bool):
+        """Resource monitoring setup.
+
+        with_bwm: Whether to start bandwidth monitor tool `bwm-ng`.  Optional because it's not available
+        in the default AL2023 distro so you have to install it first."""
+        self.mpstat_process = None
+        self.bwm_ng_process = None
+        self.with_bwm = with_bwm
+        self.output_files = []
+
+    def _start(self) -> None:
+        log.debug("Starting resource monitors...")
+        self.mpstat_process = self._start_mpstat()
+        if self.with_bwm:
+            self.bwm_ng_process = self._start_bwm_ng()
+
+    def _close(self) -> None:
+        log.debug("Shutting down resource monitors...")
+        for process in [self.mpstat_process, self.bwm_ng_process]:
+            self._stop_resource_monitor(process)
+
+        for output_file in self.output_files:
+            self._close_output_file(output_file)
+
+    def _close_output_file(self, output_file):
+        try:
+            output_file.close()
+        except Exception:
+            log.error("Error closing {output_file}:", exc_info=True)
+
+    def _stop_resource_monitor(self, process):
+        try:
+            if process:
+                process.send_signal(signal.SIGINT)
+                process.wait()
+        except Exception:
+            log.error("Error shutting down monitoring:", exc_info=True)
+
+    def _start_monitor_with_builtin_repeat(self, process_args: List[str], output_file) -> any:
+        """Start process_args with output to output_file.
+
+        Used for starting processes in the background to do monitoring; good for tools that repeat the
+        measurement themselves so only need to be started once, and that can write their output to stdout.
+        """
+        f = open(output_file, 'w')
+        self.output_files.append(f)
+        log.debug(f"Starting monitoring tool {''.join(process_args)}")
```

Comment:
I think this will output the args but with no spaces :(

```suggestion
        log.debug(f"Starting monitoring tool {' '.join(process_args)}")
```